packages/node/src/indexer/fetch.service.ts

```diff
@@ -100,41 +186,106 @@ export class FetchService implements OnApplicationShutdown {
   latestProcessed(height: number): void {
     this.latestProcessedHeight = height;
   }
-
   async startLoop(initBlockHeight: number): Promise<void> {
+    await Promise.all([
+      this.fillNextBlockBuffer(initBlockHeight),
+      this.fillBlockBuffer(),
+    ]);
+  }
+
+  async fillNextBlockBuffer(initBlockHeight: number): Promise<void> {
     if (isUndefined(this.latestProcessedHeight)) {
       this.latestProcessedHeight = initBlockHeight - 1;
     }
     await this.fetchMeta(initBlockHeight);
     // eslint-disable-next-line no-constant-condition
+    let startBlockHeight = initBlockHeight;
+
     while (!this.isShutdown) {
-      const [startBlockHeight, endBlockHeight] =
-        this.nextBlockRange(initBlockHeight) ?? [];
-      if (isUndefined(startBlockHeight)) {
+      if (this.nextBlockBuffer.size > this.bufferAllowSize) {
         await delay(1);
         continue;
       }
-      logger.info(`fetch block [${startBlockHeight}, ${endBlockHeight}]`);
-      const metadataChanged = await this.fetchMeta(endBlockHeight);
-      const blocks = await (this.nodeConfig.preferRange
-        ? SubstrateUtil.fetchBlocksViaRangeQuery(
-            this.api,
+      if (this.useDictionary) {
+        try {
+          const dictionary = await this.dictionaryService.getDictionary(
             startBlockHeight,
-            endBlockHeight,
-          )
-        : SubstrateUtil.fetchBlocks(
-            this.api,
-            startBlockHeight,
-            endBlockHeight,
-            metadataChanged ? undefined : this.parentSpecVersion,
-          ));
+            this.nodeConfig.batchSize,
+            this.project.network.dictionary,
+            this.projectIndexFilters,
+          );
+          const metaData = dictionary._metadata;
+          const batchBlocks = dictionary.batchBlocks;
+          //TODO
+          // const specVersionMap = dictionary.specVersions;
+          if (
+            metaData.chain !== this.api.runtimeChain.toString() ||
+            metaData.specName !== this.api.runtimeVersion.specName.toString() ||
+            metaData.genesisHash !== this.api.genesisHash.toString()
+          ) {
+            logger.warn(
+              `Dictionary is not matching with this network endpoint`,
+            );
+            this.useDictionary = false;
+            continue;
+          }
+          if (metaData.lastProcessedHeight < startBlockHeight) {
+            logger.warn(
+              `Dictionary indexed block is behind current indexing block height`,
+            );
+            continue;
+          }
+          if (isUndefined(batchBlocks)) {
+            await delay(1);
+            continue;
+          }
+          if (batchBlocks.length !== 0) {
+            this.nextBlockBuffer.putAll(batchBlocks);
+            this.latestBufferedHeight = batchBlocks[batchBlocks.length - 1];
+            startBlockHeight = Number(this.latestBufferedHeight) + 1;
```

Comment:
why convert it into Number here, rather than the place you assign new value to this.latestBufferedHeight